pinot-tools/src/main/java/com/linkedin/pinot/tools/realtime/provisioning/MemoryEstimator.java

```diff
@@ -0,0 +1,342 @@
+    }
+
+    RealtimeIndexOffHeapMemoryManager memoryManager = new DirectMemoryManager(_segmentMetadata.getName());
+    RealtimeSegmentZKMetadata segmentZKMetadata =
+        getRealtimeSegmentZKMetadata(_segmentMetadata, _segmentMetadata.getTotalRawDocs());
+
+    // create a config
+    RealtimeSegmentConfig.Builder realtimeSegmentConfigBuilder =
+        new RealtimeSegmentConfig.Builder().setSegmentName(_segmentMetadata.getName())
+            .setStreamName(_segmentMetadata.getTableName())
+            .setSchema(_segmentMetadata.getSchema())
+            .setCapacity(_segmentMetadata.getTotalDocs())
+            .setAvgNumMultiValues(_avgMultiValues)
+            .setNoDictionaryColumns(_noDictionaryColumns)
+            .setInvertedIndexColumns(_invertedIndexColumns)
+            .setRealtimeSegmentZKMetadata(segmentZKMetadata)
+            .setOffHeap(true)
+            .setMemoryManager(memoryManager)
+            .setStatsHistory(sampleStatsHistory);
+
+    // create mutable segment impl
+    MutableSegmentImpl mutableSegmentImpl = new MutableSegmentImpl(realtimeSegmentConfigBuilder.build());
+
+    // read all rows and index them
+    try (PinotSegmentRecordReader segmentRecordReader = new PinotSegmentRecordReader(_sampleCompletedSegment);) {
+      GenericRow row = new GenericRow();
+      while (segmentRecordReader.hasNext()) {
+        segmentRecordReader.next(row);
+        mutableSegmentImpl.index(row);
+        row.clear();
+      }
+    } catch (Exception e) {
+      throw new RuntimeException("Caught exception when indexing rows");
+    }
+
+    // dump stats into stats file
+    mutableSegmentImpl.destroy();
+
+    return statsFile;
+  }
+
+  /**
+   * Given a sample stats file, retention and total consuming partitions, estimates the amount of memory required per host
+   * @param statsFile - stats file from a sample segment for the same table
+   * @param numHosts - list of number of hosts that are to be provisioned
+   * @param numHours - list of number of hours to be consumed
+   * @param totalConsumingPartitions - total consuming partitions we are provisioning for
+   * @param retentionHours - what is the amount of retention in memory expected for completed segments
+   * @throws IOException
+   */
+  public void estimateMemoryUsed(File statsFile, int[] numHosts, int[] numHours, int totalConsumingPartitions,
+      int retentionHours) throws IOException {
+    _totalMemoryPerHost = new String[numHours.length][numHosts.length];
+    _optimalSegmentSize = new String[numHours.length][numHosts.length];
+    _consumingMemoryPerHost = new String[numHours.length][numHosts.length];
+
+    int i = 0;
+    for (int numHoursToConsume : numHours) {
+      long minutesToConsume = numHoursToConsume * 60;
+      // consuming for _numHoursSampleSegmentConsumed, gives size sampleCompletedSegmentSizeBytes
+      // hence, consuming for numHoursToConsume would give:
+      long completedSegmentSizeBytes =
```

Comment:
Do you want to do this math in seconds instead of minutes? For high ingestion topics, it may make significant difference?

Reply:
Sure, changed it to seconds. Though this would depend on whether the user of the tool is precise enough in the inputs